File: DemoApp Objective-C/CASSample/NativeAd/NativeVC.m
//
//  NativeVC.m
//  CASSample
//

#import "NativeVC.h"
#import "AppDelegate.h"

@interface NativeVC ()

@end

@implementation NativeVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self loaderSetup];
    [self registerAssetViews];
    
    self.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeNever;
    self.navigationItem.title = @"Native Ad";
}

- (void)loaderSetup {
    self.nativeLoader = [[CASNativeLoader alloc] initWithCasID: [AppDelegate casId]];
    self.nativeLoader.delegate = self;
    self.nativeLoader.adChoicesPlacement = CASChoicesPlacementTopRight;
    self.nativeLoader.isStartVideoMuted = YES;
}

- (void)registerAssetViews {
    self.nativeContainerView.headlineView = self.headlineLabel;
    self.nativeContainerView.bodyView = self.bodyLabel;
    self.nativeContainerView.iconView = self.iconView;
    self.nativeContainerView.mediaView = self.mediaView;
    self.nativeContainerView.callToActionView = self.actionButton;
}

- (IBAction)loadAction:(UIButton *)sender {
    [self.nativeLoader loadAd];
}

#pragma mark - CASImpressionDelegate

- (void)adDidRecordImpressionWithInfo:(CASContentInfo * _Nonnull)info {
    NSLog(@"adDidRecordImpressionWithInfo");
}

#pragma mark - CASNativeLoaderDelegate

- (void)nativeAdDidLoadContent:(CASNativeAdContent * _Nonnull)ad {
    ad.delegate = self;
    ad.impressionDelegate = self;
    ad.rootViewController = self;
    [self.nativeContainerView setNativeAd:ad];
}

- (void)nativeAdDidFailToLoadWithError:(CASError * _Nonnull)error {
    NSLog(@"nativeAdDidFailToLoadWithError %@", error.description);
}

#pragma mark - CASNativeAdContentDelegate

- (void)nativeAd:(CASNativeAdContent * _Nonnull)ad didFailToPresentWithError:(CASError * _Nonnull)error {
    NSLog(@"nativeAd:didFailToPresentWithError %@", error.description);
}

- (void)nativeAdDidClickContent:(CASNativeAdContent * _Nonnull)ad {
    NSLog(@"nativeAdDidClickContent");
}

@end

File: DemoApp Swift/CASSample/NativeAd/NativeVC.swift
//
//  NativeVC.swift
//  CASSample
//
//  Copyright © 2025 Clever Ads Solutions. All rights reserved.
//

import UIKit
import CleverAdsSolutions

class NativeVC: UIViewController {
            
    @IBOutlet var loadButton: UIButton!
    // Native UI
    @IBOutlet var nativeContainerView: CASNativeView!
    @IBOutlet var adStackView: UIStackView!
    @IBOutlet var adLabel: UILabel!
    @IBOutlet var bodyLabel: UILabel!
    @IBOutlet var headlineLabel: UILabel!
    @IBOutlet var iconView: UIImageView!
    @IBOutlet var mediaView: CASMediaView!
    @IBOutlet var actionButton: UIButton!
            
    private var nativeLoader = CASNativeLoader(casID: AppDelegate.casId)
        
    override func viewDidLoad() {
        super.viewDidLoad()
        loaderSetup()
        registerAssetViews()
        //
        navigationItem.largeTitleDisplayMode = .never
        navigationItem.title = "Native Ad"
    }        
        
    private func loaderSetup() {
        nativeLoader.delegate = self
        nativeLoader.adChoicesPlacement = AdChoicesPlacement.topRight // by default
        nativeLoader.isStartVideoMuted = true // by default
    }
  
    private func registerAssetViews() {
        nativeContainerView.headlineView = headlineLabel
        nativeContainerView.bodyView = bodyLabel
        nativeContainerView.iconView = iconView
        nativeContainerView.mediaView = mediaView
        nativeContainerView.callToActionView = actionButton
    }
    
    @IBAction private func loadAction(_ sender: UIButton) {
        nativeLoader.loadAd()
    }
}

extension NativeVC: CASImpressionDelegate {
    func adDidRecordImpression(info: AdContentInfo) {
        print(#function, "Ad Source: \(info.sourceName)")
    }
}

extension NativeVC: CASNativeLoaderDelegate {
    func nativeAdDidLoadContent(_ ad: NativeAdContent) {
        ad.delegate = self
        ad.impressionDelegate = self
        ad.rootViewController = self
        nativeContainerView.setNativeAd(ad)
        print(#function)
    }
  
    func nativeAdDidFailToLoad(error: AdError) {
        print(#function, "Error: \(error.description)")
    }
}

extension NativeVC: CASNativeContentDelegate {
    func nativeAd(_ ad: NativeAdContent, didFailToPresentWithError error: AdError) {
        print(#function, "Error: \(error.description)")
    }
    
    func nativeAdDidClickContent(_ ad: NativeAdContent) {
        print(#function)
    }
}
